python/rainbow/simulators/proximal_contact/gauss_seidel_solver.py

```diff
@@ -0,0 +1,124 @@
+            w += self.WJT.tocsr()[:, block].dot(delta)
+
+
+@nb.njit(parallel=True, nogil=True, cache=True)
+def sweep_worker(color_group, J, WJT, b, mu, r, x, w, delta_ws, friction_solver):
+    """ The worker function of the parallel gauss seidel algorithm.
+
+    Args:
+        color_group (ArrayLike): The color group, the value is  the block location, containing the start and end index.
+        J (ArrayLike): The contact jacobi matrix.
+        WJT (ArrayLike): The WJ^T matrix, here W = M^{-1}, and the M is the mass matrix.
+        b (ArrayLike): b = Ju^t + ∆t JM^{-1}h + EJu^t
+        mu (float): The coefficient of friction.
+        r (float): r-factor value.
+        x (ArrayLike): The current contact force.
+        w (ArrayLike): The WJT.dot(x).
+        delta_ws (ArrayLike): A array of the delta_w, each delta_w = WJT.dot(delta_x), here delta_x is the change of the contact force.
+        friction_solver (callable): The proximal operator of friction cone function.
+
+    Returns:
+        (ArrayLike, ArrayLike): The new contact force and the new delta_ws.
+    """
+    for i in nb.prange(len(color_group)):
+        block_start, block_end = color_group[i]
+        block = np.arange(block_start, block_end)
+        x_b = x[block]
+        r_b = r[block]
+        b_b = b[block]
+
+        delta = x_b.copy()
+        z_b = x_b - np.multiply(r_b, (J.dot(w)[block] + b_b))
+
+        x_b[0] = np.max(np.array([0.0, z_b[0]]))
+
+        mu_k = mu[block_start // 4]
+        x_b[1], x_b[2], x_b[3] = friction_solver(z_b[1], z_b[2], z_b[3], mu_k, x_b[0])
+
+        np.subtract(x_b, delta, delta)
+        delta_w = WJT[:, block].dot(delta)
+
+        x[block] = x_b
+        delta_ws[block_start//4] = delta_w
+
+    return x, delta_ws
+
+
+class ParallelGaussSeidelSolver(SolverInterface):
+    """ Parallel Gauss Seidel Solver.
+    """
+    def __init__(self, J, WJT, b, mu, friction_solver, engine, stats, debug_on, prefix):
+        super().__init__(J, WJT, b, mu, friction_solver, engine, stats, debug_on, prefix)
+    
+    def sweep(self):
+        # Compute the color group
+        color_groups = defaultdict(list)
+        G = BLOCKING.build_contact_graph(self.J.toarray())
+        color_groups = BLOCKING.greedy_graph_coloring(G)
+
+        # Set the number of threads
+        nb.set_num_threads(4 if (nb.config.NUMBA_NUM_THREADS // 2) >= 4 else nb.config.NUMBA_NUM_THREADS // 2)
+
+        # Compute the new contact force
+        w0 = self.WJT.dot(self.x)
+        delta_ws = np.zeros((self.WJT.shape[1]//4, w0.shape[0]), dtype=np.float64)
+        for color_group in color_groups.values():
+            w = w0.copy()
+            block_start, _ = color_group[0]
+            for i in range(block_start // 4):
+                w += delta_ws[i]
+
+            self.x, delta_ws = sweep_worker(np.array(color_group), 
+                                            self.J.toarray(), 
+                                            self.WJT.toarray(), 
```

Comment:
Ouch, these too go dense... this is very expensive. Can it not be avoided?

Reply:
Here is a trade-off using Numba because Numba can not support the spare matrix format in non-python mode.